File: datasets/python_source-code/Multiple-A1/organic5_binary_search.py
"""
Binary Search - Submission by Nathan
Approach: Half-open interval [lo, hi) instead of inclusive bounds -
hi starts at len(arr) and the loop condition is lo < hi.
"""


def locate(arr, target):
    lo = 0
    hi = len(arr)

    while lo < hi:
        mid = (lo + hi) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            lo = mid + 1
        else:
            hi = mid

    return -1


if __name__ == "__main__":
    ARR = [2, 5, 8, 12, 16, 23, 38, 45, 56, 72, 91]
    TARGET = 23
    outcome = locate(ARR, TARGET)
    print("Search outcome:", outcome)

File: server/datasets/java_source-code/Multiple-A1/organic5_binary_search.java
/*
 * Binary Search - Submission by Nathan
 * Approach: Half-open interval [lo, hi) instead of inclusive bounds -
 * hi starts at arr.length and the loop condition is lo < hi.
 */

class BinarySearch {
    public static int locate(int[] arr, int target) {
        int lo = 0;
        int hi = arr.length;

        while (lo < hi) {
            int mid = (lo + hi) / 2;
            if (arr[mid] == target) {
                return mid;
            } else if (arr[mid] < target) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }

        return -1;
    }

    public static void main(String[] args) {
        int[] arr = {2, 5, 8, 12, 16, 23, 38, 45, 56, 72, 91};
        int target = 23;
        int outcome = locate(arr, target);
        System.out.println("Search outcome: " + outcome);
    }
}
